### PR TITLE
Prometheus Metrics Support

### PR DESCRIPTION
**Pull request by Michael-Vol**
**What kind of change does this PR introduce?** 
This PR adds support for metrics collection. Each module collects a variety of default metrics (e.g. request latencies, gRPC errors, admin/client requests etc.). In addition to these defaults every module also records and collects some custom metrics regarding its function. For instance the authentication module records any login requests that have been sent.

Each module exposes a Prometheus compatible endpoint ```/metrics```, providing support for Prometheus scraping. To enable metrics, each module must be provided with a ```METRICS_PORT``` env key for the Express server port.

Metrics collection is totally optional and can be enabled to some or all of the modules.

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** 

- [ ] Yes
- [x] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `main` branch
- [ ] When resolving a specific issue, it's referenced in the PR's description  (e.g. `fix #xxx`, where "xxx" is the issue number)

### CODE REVIEW

File: libraries/grpc-sdk/src/classes/ManagedModule.ts [hunk 1]
```diff
@@ -78,6 +78,12 @@ export abstract class ManagedModule<T> extends ConduitServiceModule {
    */
   async onConfig() {}
 
+  /**
+   * This is triggered when a module needs to initialize its own custom metric
+   * types and configuration by using the sdk's ConduitMetrics?.
```
Comment (kon14) [inline libraries/grpc-sdk/src/classes/ManagedModule.ts:83]: Remove the optional chaining `?` reference

File: modules/database/src/metrics/index.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+export default {
+  databaseQueries: {
+    type: MetricType.Counter,
+    config: {
+      name: 'database_queries_total',
+      help: 'Tracks the total number of database queries',
+    },
+  },
+  databaseQueryErrors: {
+    //TODO: create global query error handler to track this
```
Comment (kon14) [inline modules/database/src/metrics/index.ts:12]: Comment this out until implemented?

File: modules/storage/src/storage-provider/providers/google/index.ts [hunk 3]
```diff
@@ -94,7 +97,16 @@ export class GoogleCloudStorage implements IStorageProvider {
   }
 
   async delete(fileName: string): Promise<boolean | Error> {
+    let fileSize = 0;
+    try {
+      fileSize = (
+        await this._storage.bucket(this._activeBucket).file(fileName).getMetadata()
+      )[0].size; //TODO: check if this is the correct way to get the file size
```
Comment (kon14) [inline modules/storage/src/storage-provider/providers/google/index.ts:104]: We need to test this.

File: modules/storage/src/storage-provider/providers/aliyun/index.ts [hunk 4]
```diff
@@ -96,13 +100,15 @@ export class AliyunStorage implements IStorageProvider {
...
+    ConduitGrpcSdk.Metrics?.increment('files_total');
+    ConduitGrpcSdk.Metrics?.increment('storage_size_bytes_total', data.byteLength);
     return true;
   }
 
   async delete(fileName: string): Promise<boolean | Error> {
     await this._ossClient.delete(fileName);
-
+    ConduitGrpcSdk.Metrics?.decrement('files_total');
+    // ConduitGrpcSdk.Metrics?.decrement('storage_size_bytes_total', fileSize); TODO: get file size from oss client
```
Comment (kon14) [inline modules/storage/src/storage-provider/providers/aliyun/index.ts:111]: We need to figure out how to get this from aliyun.

**Review by kon14 — CHANGES_REQUESTED** [review]

File: modules/database/src/adapters/sequelize-adapter/SequelizeSchema.ts [hunk 5]
```diff
@@ -256,9 +256,11 @@ export class SequelizeSchema implements SchemaAdapter<ModelCtor<any>> {
       parsedQuery = query;
     }
     if (parsedQuery.hasOwnProperty('$inc')) {
-      await this.model.increment(parsedQuery['$inc'], { where: { _id: id } }).catch(e => {
-        ConduitGrpcSdk.Logger.error(e);
-      });
+      await this.model
+        ?.increment(parsedQuery['$inc'], { where: { _id: id } })
```
[the remark this answers is not in the document]
Comment (kon14) [reply modules/database/src/adapters/sequelize-adapter/SequelizeSchema.ts]: There's another one of these in the same file.

File: modules/database/src/adapters/sequelize-adapter/SequelizeSchema.ts [hunk 6]
```diff
@@ -256,11 +265,10 @@ export class SequelizeSchema implements SchemaAdapter<ModelCtor<any>> {
       parsedQuery = query;
     }
     if (parsedQuery.hasOwnProperty('$inc')) {
-      await this.model
-        ?.increment(parsedQuery['$inc'], { where: { _id: id } })
```
Comment (kon14) [inline modules/database/src/adapters/sequelize-adapter/SequelizeSchema.ts]: There's another instance of this in the same file.

File: modules/database/src/adapters/sequelize-adapter/SequelizeSchema.ts [hunk 7]
```diff
@@ -20,6 +20,9 @@ import ConduitGrpcSdk, { ConduitSchema, Indexable } from '@conduitplatform/grpc-
 
 const deepdash = require('deepdash/standalone');
 
+const incrementDBQueries = () =>
```
Comment (kon14) [inline modules/database/src/adapters/sequelize-adapter/SequelizeSchema.ts:23]: Make this camel case.

**Review by kon14 — CHANGES_REQUESTED** [review]

**Review by kon14 — APPROVED** [review]

**Review by kkopanidis — APPROVED** [review]

**Pull request merged**